src/dbnode/encoding/proto/README.md

```diff
@@ -0,0 +1,269 @@
+
+For example, if we had the following Protobuf schema:
+
+```protobuf
+message Foo {
+	reserved 2, 3;
+	string query = 1;
+	int32 page_number = 4;
+}
+```
+
+We would begin by encoding `4` as a `varint` since that is the highest non-reserved field number.
+
+Next, we would begin encoding the field numbers and their field types 3 bits at a time where the field number is implied from their position in the list (starting at index 1 since Protobuf field numbers start at 1) and the type is encoded in the 3 bit combination:
+
+`string query = 1;` is encoded as the first value (indicating field number 1) and with the bit combination `111` indicating that it should be treated as `bytes` for compression purposes.
+
+Next, we would encode `000` twice to indicate that we will not be performing any custom compression for field numbers `2` and `3` since they are reserved.
+
+Finally, we'll encode `010` as the fourth item to indicate that field number `4` will be treated as a signed 32 bit integer.
+
+#### Custom Types
+
+0. (`000`): Not custom encoded - This type means that we won't apply any custom compression to this field and we'll rely on the standard Protobuf marshaling to encode it.
+1. (`001`): Signed 64 bit integer (`int64`, `sint64`)
+2. (`010`): Signed 32 bit integer (`int32`, `sint32`, `enum`)
+3. (`011`): Unsigned 64 bit integer (`uint64`. `fixed64`)
+4. (`100`): Unsigned 32 bit integer (`uint32`, `fixed32`)
+5. (`101`): 64 bit float (`double`)
+6. (`110`): 32 bit float (`float`)
+7. (`111`): bytes (`bytes`, `string`)
+
+### Stream of Writes
+
+After encoding the header, the stream becomes a list of tuples in the form of `<control bit, compressed timestamp, compressed custom encoded fields, protobuf marshaled fields>`
+
+#### Control Bit
+
+TODO(rartoul): This will change slightly once we implement support for mid-stream scheme changes, we'll probably add one more optional control bit in the case where the initial control bit is zero which will have minimal impact on compressions since it will only be present when we do a schema change, a time unit change or the stream ends.
+
+Every write is prefixed with a control bit.
+
+If the control bit is set to `1` then we know that the stream contains another write that we need to decode and we can begin decoding the timestamp.
+
+If the control bit is set to `0` then we have either: **a)** reached the end of the stream *or* **b)** encountered a time unit change.
+
+We resolve the ambiguity by reading the next control bit which will be `0` if we've reached the end of the stream or `1` if we should interpret the next 8 bits as the new time unit that we should use for delta of delta timestamp encoding.
+
+We have to keep track of time unit changes on our own instead of deferring to the M3TSZ delta-of-delta encoder which can handle this independently because the M3TSZ encoder relies on a custom marker scheme to indicate time unit changes.
```

Comment:
> Time unit changes must be tracked manually (instead of deferring to the M3TSZ delta-of-delta encoder, which can handle this independently) because the M3TSZ encoder relies [...]